### ISSUE TITLE
narrowing to "never" by "if" statement does not exclude "undefined" return type

### ISSUE THREAD

**Issue by babatakao** [Microsoft/TypeScript#21985]
**TypeScript Version:**  2.7.1

**Search Terms:**  narrowing types never switch case if undefined strictNullChecks

**Code**

```ts
// Build this code with "--strictNullChecks" option.

function assertNever(x: never): never {
    throw new Error("not reached");
}

enum A {
    Foo, Bar
}

// There are no errors on this function.
function good(a: A): number {
    switch (a) {
        case A.Foo: return 0;
        case A.Bar: return 1;
    }
}

// This function causes the following error though "a" is "never" at the end of the function.
// error TS2366: Function lacks ending return statement and return type does not include 'undefined'.
function bad(a: A): number {
    if (a === A.Foo) return 0;
    if (a === A.Bar) return 1;
}

// No error if explicit return of "never" is added.
function ok(a: A): number {
    if (a === A.Foo) return 0;
    if (a === A.Bar) return 1;
    return assertNever(a);
}
```

**Expected behavior:** The `bad()` function should not cause error so that `if` statement works like `switch`.

**Actual behavior:** As commented in the code, only `bad()` function causes an error.

**Playground Link:** (please check "strictNullChecks" option.)

http://www.typescriptlang.org/play/#src=function%20assertNever(x%3A%20never)%3A%20never%20%7B%0D%0A%20%20%20%20throw%20new%20Error(%22unexpected%22)%3B%0D%0A%7D%0D%0A%0D%0Aenum%20A%20%7B%0D%0A%20%20%20%20Foo%2C%20Bar%0D%0A%7D%0D%0A%0D%0Afunction%20good(a%3A%20A)%3A%20number%20%7B%0D%0A%20%20%20%20switch%20(a)%20%7B%0D%0A%20%20%20%20%20%20%20%20case%20A.Foo%3A%20return%200%3B%0D%0A%20%20%20%20%20%20%20%20case%20A.Bar%3A%20return%201%3B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Afunction%20bad(a%3A%20A)%3A%20number%20%7B%0D%0A%20%20%20%20if%20(a%20%3D%3D%3D%20A.Foo)%20return%200%3B%0D%0A%20%20%20%20if%20(a%20%3D%3D%3D%20A.Bar)%20return%201%3B%0D%0A%7D%0D%0A%0D%0Afunction%20ok(a%3A%20A)%3A%20number%20%7B%0D%0A%20%20%20%20if%20(a%20%3D%3D%3D%20A.Foo)%20return%200%3B%0D%0A%20%20%20%20if%20(a%20%3D%3D%3D%20A.Bar)%20return%201%3B%0D%0A%20%20%20%20return%20assertNever(a)%3B%0D%0A%7D

**Related Issues:**

**Comment by jack-williams** [comment]
The issue I believe is that `switch` statements have a notion of being possibly exhaustive and in the case of `good` the switch is exhaustive so there is no implicit return at the end. I'm not sure this is directly a problem with `strictNullChecks`.

I think to fix this, `if` statements without `else` would need a notion of being possibly exhaustive too.

**Comment by VanTanev** [comment]
It seems that this isn't a high priority, but it would be a nice little quality of life addition.

For now, the `assertNever()` approach works as a stopgap.

**Comment by thitemple** [comment]
I'm facing the same issue found here, I disagree it's just a cosmetic addition if I'm relying on TS to exhaustive check using an `if`. The `assertNever` prevents TS from checking that.

```ts
type Ok = {
  ok: true,
  result: { name: string }
}

type Bad = {
  ok: false,
  error: 'NOT_FOUND'
}

function getData(x: boolean): Ok | Bad {
    if (x) {
        return { ok: true, result: { name: 'James' } }
    }
    return { ok: false, error: 'NOT_FOUND' }
}

function resolveDataIf(x: boolean): string | null {
    const result = getData(x);
    if (!result.ok) {
        if (result.error === 'NOT_FOUND') {
            return null;
        }
    }
    return result.result.name;
}

function resolveDataSwitch(x: boolean): string | null {
    const result = getData(x);
    if (!result.ok) {
        switch(result.error) {
            case 'NOT_FOUND':
                return null;
        }
    }
    return result.result.name;
}
```
The above is a simplified scenario we have. The `switch` statement works, but not the `if`. In my real code, the `error` attribute could be updated to accept values other than `NOT_FOUND`, and I'd like the consumers of that type to handle all possible types of error.

When we have only one possible value for `error` we try to stay away from switch statements just because it reads better, but then that doesn't work anymore.
